rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <chrono>
+#include <thread>
+
+#include <rmf_traffic_ros2/schedule/internal_MonitorNode.hpp>
+#include <rmf_traffic/schedule/Database.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+
+using namespace std::chrono_literals;
+
+class MissingQueryMonitorNode : public rmf_traffic_ros2::schedule::MonitorNode
+{
+public:
+  MissingQueryMonitorNode(
+    std::function<void(std::shared_ptr<rclcpp::Node>)> callback,
+    const rclcpp::NodeOptions& options)
+    : MonitorNode(callback, options, MonitorNode::no_automatic_setup)
+  {}
+
+  void announce_fail_over()
+  {
+    if (rclcpp::ok())
+    {
+      // Wait a while before announcing. This will ensure that mirrors pick up
+      // on the missing query via the query check logic, not the failover
+      // handling logic. (But the announcement still has to be made to ensure
+      // mirrors reconnect services.)
```

Comment:
Is this testing the potential issue that I mentioned [here](https://github.com/open-rmf/rmf_ros2/pull/86/files#r674955667)? If so, I'm surprised that we haven't seen that be an issue.

Is it possible that doing `std::this_thread::sleep_for` here is actually blocking the new schedule node from publishing its query update because this is all running on a single-threaded executor? So ultimately this `FailOverEvent` message does get published before the query update; it just all gets published after a 5s delay.

I'll try to run these test nodes and echo the topics to see what I get.

Reply:
Okay after testing this out manually, I understand why it's not actually triggering the edge case that I linked to above. Since the newly spawned schedule node is initially missing the query, it doesn't immediately start publishing to that query topic. Instead it publishes an empty query list, which triggers the mirror managers to re-register. However, their re-registration is held up because the `sleep_for` blocks the new schedule node from responding to the service call.

Here's where things get *wild*: Once the sleep is over, the registration requests start making it through, but the mirror manager of the monitor node destructs, which unregisters *its* query. Since it has the same query as the other mirrors that are trying to re-register, the subscriber count that we're maintaining actually drops to zero, so the new schedule node creates and then immediately discards that query, even though there are mirror managers that still want it.

At this point, query registration requests start bouncing around like crazy and the "subscription count" spikes up to ~40-60 due to massive duplication of requests from the *two* mirror managers that actually exist. The system does eventually stabilize (within about one second) and everything goes on working, but the subscription count is totally fubar and will never be able to go back down to zero.

To reproduce the issue I'm describing, I recommend running the `office.launch.xml` demo from `rmf_demos` but run the `missing_query_monitor_node` instead of the usual `rmf_traffic_schedule_monitor`. Then run these in separate terminals:
* `$ ros2 topic echo /rmf_traffic/registered_queries`
* `$ ros2 topic echo /rmf_traffic/fail_over_event`
* [`$ ros2 launch rmf_demos office_loop.launch.xml`](https://github.com/open-rmf/rmf_demos/pull/66)
* `sudo killall rmf_traffic_schedule`

You'll see some crazy output from `/rmf_traffic/registered_queries` once the message on `/rmf_traffic/fail_over_event` pops out.

I have some fixes to make the behavior a lot more sane, but I had to do a lot of debugging, so I'll need to clean that up before I push it. I should be able to do that on Monday.

An important change that we'll need is to outright get rid of the manual subscription counting. I think the whole idea is fundamentally broken in this kind of distributed system where we simply can't really know the lifecycles of the remote listeners.

An ideal alternative might be if we can give the publisher a callback that gets triggered when the number of topic subscribers changes; then we could drop the query when subscribers reach zero (and remain at zero for some time window). But it doesn't look to me like ROS2 offers that feature, so we could just take a brute force approach of doing a periodic check (every X minutes) to see if any topics have zero subscribers, and then prune them.

Reply:
Addressed by #87